test-suite/unit-tests/src/utest.ml

```diff
@@ -0,0 +1,68 @@
+open OUnit
+
+(* general case to build a test *)
+let mk_test nm test = nm >: test
+
+(* common cases for building tests *)
+let mk_eq_test nm descr actual expected =
+  mk_test nm (TestCase (fun _ -> assert_equal ~msg:descr expected actual))
+
+let mk_bool_test nm descr actual =
+  mk_test nm (TestCase (fun _ -> assert_bool descr actual))
+
+(* given test result, print message, return success boolean *)
+let logger out_ch result =
+  match result with
+  | RSuccess path ->
+     let _ = Printf.fprintf out_ch "TEST SUCCEEDED: %s\n" (string_of_path path) in
+     true
+  | RError (path,msg)
+  | RFailure (path,msg) ->
+     let _ = Printf.fprintf out_ch "TEST FAILED: %s (%s)\n" (string_of_path path) msg in
+     false
+  | RSkip (path,msg)
+  | RTodo (path,msg) ->
+     let _ = Printf.fprintf out_ch "TEST DID NOT SUCCEED: %s (%s)\n" (string_of_path path) msg in
+     false
+
+(* run one OUnit test case, return successes, no. of tests *)
+(* notionally one test, which might be a TestList *)
+let run_one logit test =
+  let rec process_results rs =
+    match rs with
+      [] -> (0,0)
+    | (r::rest) ->
+       let succ = if logit r then 1 else 0 in
+       let succ_results,tot_results = process_results rest in
+       (succ + succ_results,tot_results + 1)
+  in
+  let results = perform_test (fun _ -> ()) test in
+  process_results results
+
+(* run list of OUnit test cases, log results *)
+let run_tests ml_fn tests =
+  let log_fn = ml_fn ^ ".log" in
+  let out_ch = open_out log_fn in
+  let logit = logger out_ch in
+  let rec run_some tests succ tot =
+    match tests with
+      [] -> (succ,tot)
+    | (t::ts) ->
+       let succ_one,tot_one = run_one logit t in
+       run_some ts (succ + succ_one) (tot + tot_one)
+  in
+  (* format for test-suite summary to find status
+     success if all tests succeeded, else failure
+   *)
+  let succ,tot = run_some tests 0 0 in
+  let _ = Printf.fprintf out_ch
+            "*** Ran %d tests, with %d successes and %d failures ***\n"
+            tot succ (tot - succ)
+  in
+  if succ = tot then
+    Printf.fprintf out_ch
+      "==========> SUCCESS <==========\n    %s...Ok\n" ml_fn
+  else
+    Printf.fprintf out_ch
+      "==========> FAILURE <==========\n    %s...Error!\n" ml_fn;
```

Comment:
Failing tests should print `FAILED    foo.log` to stderr as they run, inserting `Printf.eprintf "FAILED    %s.log\n" ml_fn` in this `else` branch should work.
(github display seems to eat the spaces, for alignment use 4 spaces between FAILED and the filename)

Reply:
Please use `Format` and define a proper wrapper with flush.